File: Leetcode Daily Challenge/026-August-2025/028-sort-matrix-by-diagonals.java
// https://leetcode.com/problems/sort-matrix-by-diagonals/




import java.util.*;

class Solution {
    private int m, n;

    public int[][] sortMatrix(int[][] grid) {
        m = grid.length;
        n = grid[0].length;

        // return approach_1(grid);
        return approach_2(grid);
    }

    // TC : O(M * N)
    // SC : O(K) where K = max-length of diagonal
    private int[][] approach_2(int[][] grid) {
        for (int i = 0; i < m; i ++) {
            sortDiagonal(i, 0, grid, true);
        }
        for (int j = 1; j < n; j ++) {
            sortDiagonal(0, j, grid, false);
        }
        return grid;
    }

    void sortDiagonal(int r, int c, int[][] grid, boolean isLowerDiag) {
        List<Integer> diagonal = new ArrayList<>();
        int i = r, j = c;

        while (i < m && j < n) {
            diagonal.add(grid[i][j]);
            i ++;
            j ++;
        }

        if (isLowerDiag) {
            diagonal.sort(Collections.reverseOrder());
        } else {
            Collections.sort(diagonal);
        }

        i = r;
        j = c;
        for (int idx = 0; idx < diagonal.size(); idx ++) {
            grid[i ++][j ++] = diagonal.get(idx);
        }
    }

    // TC : O((M * N) + (K*log(K))) Where K = No. of diagonals
    // SC : O(M * N)
    private int[][] approach_1(int[][] grid) {
        Map<Integer, List<Integer>> mpp = new HashMap<>();
        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j ++) {
                int key = (i - j);
                mpp.putIfAbsent(key, new ArrayList<>());
                mpp.get(key).add(grid[i][j]);
            }
        }

        for (Map.Entry<Integer, List<Integer>> entry : mpp.entrySet()) {
            if (entry.getKey() >= 0) {
                Collections.sort(entry.getValue());
            } else {
                entry.getValue().sort(Collections.reverseOrder());
            }
        }

        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j ++) {
                List<Integer> list = mpp.get(i - j);
                grid[i][j] = list.get(list.size() - 1);
                list.remove(list.size() - 1);
            }
        }
        return grid;
    }
}

File: Leetcode Daily Challenge/026-August-2025/028-sort-matrix-by-diagonals.cpp
// https://leetcode.com/problems/sort-matrix-by-diagonals/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    // TC : O(max(m, n)  + klog(k)) ~= O(klogk) where K = max no. of elements in a diagonal
    // SC : O(k)
    void sortDiag(int r, int c, int m, int n, vector<vector<int>>& grid, bool isLowerDiag) {
        vector<int> diagElements;
        int i = r, j = c;
        while (i < m && j < n) {
            diagElements.push_back(grid[i][j]);
            i ++, j ++;
        }
        if (isLowerDiag) {
            sort(begin(diagElements), end(diagElements));
        } else {
            sort(rbegin(diagElements), rend(diagElements));
        }

        // insert into grid the sorted values
        i = r, j = c;
        while (i < m && j < n) {
            grid[i][j] = diagElements.back();
            diagElements.pop_back();
            i ++, j ++;
        }
    }

    // TC : O(M * N + M*Klog(K)) ~= O(M * N)
    // SC : O(M * N)
    vector<vector<int>> approach_1(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        unordered_map<int, vector<int>> mpp;
        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j ++) {
                mpp[i - j].push_back(grid[i][j]);
            }
        }
        for (auto& it : mpp) {
            if (it.first >= 0) 
                sort(begin(it.second), end(it.second));
            else 
                sort(rbegin(it.second), rend(it.second));
        }
        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j ++) {
                grid[i][j] = mpp[i - j].back();
                mpp[i - j].pop_back();
            }   
        }
        return grid;
    }

    // TC : O( (M + N) * Klog(K) )
    // SC : O(K)
    vector<vector<int>> approach_2(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        for (int i = 0; i < m; i ++) {
            sortDiag(i, 0, m, n, grid, 1);  
        }
        for (int j = 1; j < n; j ++) {
            sortDiag(0, j, m, n, grid, 0);  
        }
        return grid;
    }
public:
    vector<vector<int>> sortMatrix(vector<vector<int>>& grid) {
        // return approach_1(grid);     // USING HASH MAP : {int, {int}}
        return approach_2(grid);     // More of SIMULATION
    }
};